### PR TITLE
Add budget harmonization method

### PR DESCRIPTION
**Pull request by coroa**
~Includes changes from #27.~

~So to focus on new additions, it'd be good to only look at the few last commits.~

The `ipopt` solver has not been added as an explicit python dependency, since it is relatively heavy and
only necessary when one explicitly chooses the budget method. One could add it as python variant, like `test` or `doc`, though.

### CODE REVIEW

**Comment by gidden** [comment]
@coroa on this one, could you please rebase to only the commits related? I may have a few comments which would be easier to walk through as such.

**Comment by coroa** [comment]
> @coroa on this one, could you please rebase to only the commits related? I may have a few comments which would be easier to walk through as such.

Done

File: aneris/methods.py [hunk 1]
```diff
@@ -184,6 +186,131 @@ def reduce_ratio(df, ratios, final_year='2050', harmonize_year='2015'):
...
+    harmonized = []
+
+    for region in df.index:
+
+        # Step 2: Set up the optimisation problem
+
+        # Define model
+        model = pyo.ConcreteModel()
+
+        """
```
Comment (gidden) [inline aneris/methods.py:244]: Can this doc be removed?

**Review by gidden — CHANGES_REQUESTED** [review]
Thanks again @coroa - I have two doc requests, let me know if you think they are too much:

1. in the docstring for `budget` at minimum could you add a description of how this works (i.e., what is being minimized, constrained, etc.) and at maximum add the mathematical formulation?
2. we should also have already a tutorial notebook, could you port that here since this method is less 'transparent' and could very much use a visual explanation?

**Comment by coroa** [comment]
@gidden Please check whether the last commits, fulfill your requirements.

I have not used math in numpy docstrings before, I am not sure, whether sphinx will run fine with them. And a quick test reveal they are not working as i had expected. Should just have looked up docs before (https://numpydoc.readthedocs.io/en/latest/format.html). Anyway, feel free to suggest or adapt.

Otherwise i'll finish it on monday.

**Comment by gidden** [comment]
Looks great, thanks @coroa !

**Review by gidden — APPROVED** [review]

**Pull request merged**